libs/datamodel/core/src/transform/ast_to_dml/validation_pipeline/validations/relations.rs

```diff
@@ -411,6 +415,62 @@ fn cascade_error_with_default_values(relation: CompleteInlineRelationWalker<'_,
     DatamodelError::new_validation_error(&msg, relation.referencing_field().ast_field().span)
 }
 
+/// The types of the referencing and referenced scalar fields in a relation must be compatible.
+pub(super) fn referencing_scalar_field_types(relation: InlineRelationWalker<'_, '_>, diagnostics: &mut Diagnostics) {
+    let datasource = relation.db().datasource();
+    // see https://github.com/prisma/prisma/issues/10105
+    if datasource
+        .as_ref()
+        .map(|source| source.provider == MONGODB_SOURCE_NAME)
+        .unwrap_or(false)
+    {
+        return;
+    }
+
+    let referencing_fields = match relation.referencing_fields() {
+        ReferencingFields::Concrete(fields) => fields,
+        _ => return,
+    };
+
+    for (referencing, referenced) in referencing_fields.zip(relation.referenced_fields()) {
+        if !field_types_match(
+            referencing.scalar_field.r#type,
+            referenced.scalar_field.r#type,
+            relation.db(),
+        ) {
+            diagnostics.push_error(DatamodelError::new_attribute_validation_error(
+                &format!(
+                    "The type of the field `{}` in the model `{}` is not matching the type of the referenced field `{}` in model `{}`.",
+                    referencing.name(),
+                    referencing.model().name(),
+                    referenced.name(),
+                    referenced.model().name(),
+                ),
+                RELATION_ATTRIBUTE_NAME,
+                relation.forward_relation_field().unwrap().ast_field().span,
+            ))
+        }
+    }
+
+    fn field_types_match(referencing: ScalarFieldType, referenced: ScalarFieldType, db: &ParserDatabase<'_>) -> bool {
```

Comment:
Interesting. Is there a good reasoning why we have a separate function and do not match directly in the `for` loop?

Reply:
Because we need recursion for type aliases.